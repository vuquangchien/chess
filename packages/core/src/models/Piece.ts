import {PieceColor} from './PieceColor';
import {PieceType} from './PieceType';
export type FenName =
  | 'p'
  | 'r'
  | 'n'
  | 'b'
  | 'q'
  | 'k'
  | 'P'
  | 'R'
  | 'N'
  | 'B'
  | 'Q'
  | 'K';
export const PieceTypes: {[key: string]: PieceType} = {
  PAWN: 'pawn',
  ROOK: 'rook',
  KNIGHT: 'knight',
  BISHOP: 'bishop',
  QUEEN: 'queen',
  KING: 'king',
};
export const PieceColors: {[key: string]: PieceColor} = {
  WHITE: 'white',
  BLACK: 'black',
};
export class Piece {
  public color: PieceColor;
  public type: PieceType;
  private isMoved = false;

  constructor(type: PieceType, color: PieceColor) {
    this.color = color;
    this.type = type;
  }

  static fromName(name: FenName): Piece {
    const color =
      name === name.toUpperCase() ? PieceColors.WHITE : PieceColors.BLACK;
    const type = name.toLowerCase();
    const pieceType = {
      p: PieceTypes.PAWN,
      r: PieceTypes.ROOK,
      n: PieceTypes.KNIGHT,
      b: PieceTypes.BISHOP,
      q: PieceTypes.QUEEN,
      k: PieceTypes.KING,
    }[type];
    if (!pieceType) throw new Error(`Unknown piece type: ${type}`);
    return new Piece(pieceType, color);
  }

  public getColor(): PieceColor {
    return this.color;
  }

  public isPieceMoved(): boolean {
    return this.isMoved;
  }
  public setPieceMoved(): void {
    this.isMoved = true;
  }
  public getType(): PieceType {
    return this.type;
  }
  public getIcon(): string {
    const icons = {
      rook: {
        white: '♖',
        black: '♜',
      },
      knight: {
        white: '♘',
        black: '♞',
      },
      bishop: {
        white: '♗',
        black: '♝',
      },
      queen: {
        white: '♕',
        black: '♛',
      },
      king: {
        white: '♔',
        black: '♚',
      },
      pawn: {
        white: '♙',
        black: '♟',
      },
    };
    return icons[this.type][this.color];
  }
}
