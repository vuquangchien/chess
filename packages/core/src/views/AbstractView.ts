import {IView} from '../framework/IView';
import {IEngine} from './IEngine';

export abstract class AbstractView<T> implements IView<T> {
  engine!: IEngine;
  children: AbstractView<unknown>[] = [];
  controller: T;

  public constructor(controller: T) {
    this.children = [];
    this.controller = controller;
    this.init();
  }
  init(): void {}

  setEngine(engine: IEngine): void {
    this.engine = engine;
    this.children.forEach(child => child.setEngine(engine));
  }
  addChild(child: AbstractView<unknown>): void {
    this.children.push(child);
  }

  addChildren(children: AbstractView<unknown>[]): void {
    this.children.push(...children);
  }

  renderChildren(): string {
    let result = '';
    for (const child of this.children) {
      result += child.render();
    }
    return result;
  }

  abstract render(): string;
}
