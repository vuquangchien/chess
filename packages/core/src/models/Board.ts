import {Cell} from './Cell';
import {CellIterator} from './CellIterator';
import * as _ from 'lodash';
import {Move} from './moves/Move';
import {MoveFactory} from './moves/MoveFactory';
import {PieceColors} from './Piece';
import {PieceColor} from './PieceColor';
import {PolicyFactory} from './policies/PolicyFactory';

export class Board {
  public cells: Cell[][];
  public moves: Move[] = [];
  private player = PieceColors.WHITE;
  constructor() {
    // const cellIndex = 0;
    const files: Cell[][] = [];
    for (let rowIndex = 0; rowIndex < 8; rowIndex++) {
      const file: Cell[] = [];
      for (let colIndex = 0; colIndex < 8; colIndex++) {
        const cell = new Cell(colIndex, rowIndex, null);
        file.push(cell);
      }
      files.push(file);
    }
    this.cells = files;
  }

  public getByCellName(name: string): Cell {
    try {
      const file = name.charCodeAt(0) - 97;
      const rank = 8 - parseInt(name.charAt(1), 10);
      return this.cells[rank][file];
    } catch (e) {
      throw new Error(`Invalid cell name ${name}`);
    }
  }

  public assertMove(move: Move): void {
    const movePolicies = PolicyFactory.getPolicies(this, move);
    movePolicies.forEach(policy => {
      const error = policy.isMoveValid(move);
      if (error) {
        throw new Error(error);
      }
    });
  }

  public movePiece(from: Cell, to: Cell): void {
    const move = MoveFactory.createMove(from, to);
    this.assertMove(move);
    this.moves.push(move);
    to.setPiece(from.piece);
    from.piece!.setPieceMoved();
    from.setPiece(null);
    this.player =
      this.player === PieceColors.WHITE ? PieceColors.BLACK : PieceColors.WHITE;
  }
  public movePieceByCellName(from: string, to: string): void {
    const fromCell = this.getByCellName(from);
    const toCell = this.getByCellName(to);
    this.movePiece(fromCell, toCell);
  }
  public getCellIterator(): CellIterator {
    return new CellIterator(_.flatten(this.cells));
  }
  public highlightPossibleMoves(cell: Cell) {
    const possibleMoves = this.getPossibleMoves(cell);
    possibleMoves.forEach(possibleMove => {
      possibleMove.highlight();
    });
  }
  public getPossibleMoves(fromCell: Cell): Cell[] {
    const possibleMoves: Cell[] = [];
    const it = this.getCellIterator();
    while (it.hasNext()) {
      const toCell = it.next();
      try {
        const move = MoveFactory.createMove(fromCell, toCell);
        this.assertMove(move);
        possibleMoves.push(toCell);
      } catch (e) {
        // console.log(e);
      }
    }
    return possibleMoves;
  }
  public getAllPossibleMoves(): Move[] {
    const possibleMoves: Move[] = [];
    const it = this.getCellIterator();
    while (it.hasNext()) {
      const fromCell = it.next();
      if (!fromCell.piece) {
        continue;
      }
      const moves = this.getPossibleMoves(fromCell);
      moves.forEach(toCell => {
        possibleMoves.push(MoveFactory.createMove(fromCell, toCell));
      });
    }
    return possibleMoves;
  }
  resetHighlight() {
    const it = this.getCellIterator();
    while (it.hasNext()) {
      it.next().unHighlight();
    }
  }

  getCurrentPlayer() {
    return this.player;
  }

  setCurrentPlayer(player: PieceColor) {
    this.player = player;
  }

  clone() {
    return _.cloneDeep(this);
  }
}
