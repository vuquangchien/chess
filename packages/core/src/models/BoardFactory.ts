import {Board} from './Board';
import {FenRecord} from './fen/FenRecord';

export class BoardFactory {
  public static createEmptyBoard(): Board {
    return new Board();
  }
  public static createFromFen(fen: string): Board {
    const board = new Board();
    const fenRecord = new FenRecord(fen);
    const cells = fenRecord.piecePlacement.getCells();
    cells.forEach(cell => {
      board.getByCellName(cell.name).setPiece(cell.piece);
    });
    board.setCurrentPlayer(fenRecord.activeColor);
    return board;
  }
  public static createBoardWithPieces(): Board {
    return BoardFactory.createFromFen(
      'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w - - 0 1'
    );
  }
}
