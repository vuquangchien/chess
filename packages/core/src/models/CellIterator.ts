import {Board} from './Board';
import {Cell} from './Cell';
import {Rank} from './Rank';

export class RankIterator {
  private board: Board;
  private index: number;

  constructor(board: Board) {
    this.board = board;
    this.index = -1;
  }

  hasNext(): boolean {
    return this.index < 7;
  }

  currentIndex(): number {
    return this.index;
  }

  next(): Rank {
    this.index++;
    return this.current();
  }

  private current(): Rank {
    return this.board.getRanks()[this.index];
  }
}
export class CellIterator {
  private cells: Cell[];
  private index: number;

  constructor(cells: Cell[]) {
    this.cells = cells;
    this.index = -1;
  }

  hasNext(): boolean {
    return this.index < this.cells.length - 1;
  }

  next(): Cell {
    this.index++;
    return this.current();
  }

  private current(): Cell {
    return this.cells[this.index];
  }
}
