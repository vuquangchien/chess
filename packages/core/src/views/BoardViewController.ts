import {Board} from '../models/Board';

export class BoardViewController {
  private board: Board;

  constructor(board: Board) {
    this.board = board;
  }

  public getBoard(): Board {
    return this.board;
  }
}
