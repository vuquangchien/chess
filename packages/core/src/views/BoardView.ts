import {CellView} from './CellView';
import {CellViewController} from './CellViewController';

import {BoardViewController} from './BoardViewController';
import {RowEl, TextEl} from './RowEl';
import {AbstractView} from './AbstractView';

export class BoardView extends AbstractView<BoardViewController> {
  public init(): void {
    this.addChild(new RowEl());
    const headerRow = new RowEl();
    headerRow.addChildren(
      '   a b c d e f g h'
        .split(' ')
        .map(el => new TextEl({text: el + '', row: 0, column: 0}))
    );
    this.addChild(headerRow);
    const board = this.controller.getBoard();
    const iterator = board.getRankIterator();
    while (iterator.hasNext()) {
      const row = new RowEl();
      row.addChild(new TextEl({text: `${7 - iterator.currentIndex()} `}));
      const rank = iterator.next();
      const it = rank.getIterator();
      while (it.hasNext()) {
        const cell = it.next();
        const vc = new CellViewController(cell);
        const cellView = new CellView(vc);
        row.addChild(cellView);
      }
      this.addChild(row);
    }
  }

  public render(): string {
    return this.renderChildren();
  }
}
