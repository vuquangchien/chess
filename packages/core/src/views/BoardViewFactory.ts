import {Board} from '../models/Board';
import {BoardViewController} from './BoardViewController';
import {BoardView} from './BoardView';

export class BoardViewFactory {
  public static fromBoard(board: Board): BoardView {
    return new BoardView(new BoardViewController(board));
  }
}
