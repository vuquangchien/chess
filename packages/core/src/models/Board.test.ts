import {Board} from './Board';
// @ts-ignore
import {beforeEach, expect} from '@jest/globals';
import {BoardFactory} from './BoardFactory';
import {BoardView} from '../views/BoardView';
import {BoardViewController} from '../views/BoardViewController';
import {Piece, PieceColors, PieceTypes} from './Piece';
import {RankIterator} from './CellIterator';
import {BoardViewFactory} from '../views/BoardViewFactory';

describe('Board', () => {
  let board: Board;
  beforeEach(() => {
    board = BoardFactory.createBoardWithPieces();
  });
  it('should create empty board correctly', () => {
    const board = BoardFactory.createEmptyBoard();
    const cells = board.cells;
    expect(cells[0][0].name).toEqual('a8');
    expect(cells[0][7].name).toEqual('h8');
    expect(cells[7][0].name).toEqual('a1');
    expect(cells[7][7].name).toEqual('h1');
    expect(cells[0][0].getCellIndex()).toEqual(0);
    expect(cells[0][7].getCellIndex()).toEqual(7);
    expect(cells[7][0].getCellIndex()).toEqual(56);
    expect(cells[7][7].getCellIndex()).toEqual(63);
  });
  it('should set piece', () => {
    const board = new Board();
    const piece = new Piece(PieceTypes.ROOK, PieceColors.WHITE);
    board.cells[0][0].setPiece(piece);
    const cell = board.cells[0][0];
    expect(cell.piece).toEqual(piece);
  });

  it('should create new board', () => {
    const board = new Board();
    expect(board).toBeDefined();
    expect(board.cells.length).toEqual(8);
    const firstCell = board.cells[0][0];
    const lastCell = board.cells[7][7];
    expect(firstCell.name).toEqual('a8');
    expect(lastCell.name).toEqual('h1');
  });
  it('should get ranks', () => {
    const ranks = new RankIterator(new Board()).ranks;
    expect(ranks.length).toEqual(8);
    const firstRank = ranks[0];
    expect(firstRank.cells[0].name).toEqual('a8');
    expect(firstRank.cells[7].name).toEqual('h8');
  });
  it('should iterate cells', () => {
    const board = new Board();
    const it = new RankIterator(board);
    while (it.hasNext()) {
      const rank = it.next();
      const cellIterator = rank.getIterator();
      while (cellIterator.hasNext()) {
        const cell = cellIterator.next();
        expect(cell).toBeDefined();
      }
    }
  });
  it('should allow get by cell name', () => {
    const board = new Board();
    const cell = board.getByCellName('a8');
    expect(cell.name).toEqual('a8');
  });
  it('should allow to move piece', () => {
    const board = BoardFactory.createBoardWithPieces();
    board.setCurrentPlayer(PieceColors.BLACK);
    const cell = board.getByCellName('a7');
    board.movePiece(cell, board.getByCellName('a6'));
    expect(cell.piece).toBeNull();
    expect(board.getByCellName('a6').piece).toBeDefined();
    expect(board.moves.length).toEqual(1);
    expect(board.moves[0].from.name).toEqual('a7');
    expect(board.moves[0].to.name).toEqual('a6');
    expect(board.moves[0].player).toEqual(PieceColors.BLACK);

    expect(
      new BoardView(new BoardViewController(board)).render()
    ).toMatchSnapshot();
  });
  it('should mark piece as moved', () => {
    const board = BoardFactory.createBoardWithPieces();
    const cell = board.getByCellName('a2');
    expect(cell.piece).toBeDefined();
    expect(cell.piece!.isPieceMoved()).toBeFalsy();
    board.movePiece(cell, board.getByCellName('a3'));
    expect(board.getByCellName('a3').piece!.isPieceMoved()).toBeTruthy();
  });
  it('should flip current player after move', () => {
    const board = BoardFactory.createBoardWithPieces();
    expect(board.getCurrentPlayer()).toEqual(PieceColors.WHITE);
    board.movePieceByCellName('a2', 'a3');
    expect(board.getCurrentPlayer()).toEqual(PieceColors.BLACK);
  });
  it('should clone board', () => {
    const board = BoardFactory.createBoardWithPieces();
    const clone = board.clone();
    board.movePieceByCellName('b1', 'c3');
    expect(clone.getByCellName('b1').piece).toBeDefined();
    expect(BoardViewFactory.fromBoard(clone).render()).not.toEqual(
      BoardViewFactory.fromBoard(board).render()
    );
  });
});
