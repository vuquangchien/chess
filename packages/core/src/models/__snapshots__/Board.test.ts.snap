// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should allow to move piece 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □♜□♞□♝□♛□♚□♝□♞□♜
□7 □ □♟□♟□♟□♟□♟□♟□♟
□6 □♟□ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □ □ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □♙□♙□♙□♙□♙□♙□♙□♙
□1 □♖□♘□♗□♕□♔□♗□♘□♖
"
`;
